src/generic.jl

```diff
@@ -109,17 +109,48 @@ function _pairwise!(r::AbstractMatrix, metric::SemiMetric, a::AbstractMatrix)
     size(r) == (n, n) || throw(DimensionMismatch("Incorrect size of r."))
     @inbounds for j = 1:n
         aj = view(a, :, j)
+        for i = 1:(j - 1)
+            r[i, j] = r[j, i]   # leveraging the symmetry of SemiMetric
+        end
```

Comment:
There are two things I don't understand:
- How did this work before? Don't we already do this somewhere else?
- Why copy *before* filling the other part?

Reply:
How it worked before: let's say, we want to fill a 3x3 matrix symmetrically, then the old write pattern was like this:
```
 1  6  8
 2  4  9
 3  5  7
```
In 1, 4, 7 we simply fill with a zero. In 2, 3, 5 we compute new distances, in 6, 8 and 9 we copy the values from 2, 3 and 5, respectively, i.e., values computed in earlier iterations.

How it's working here:
```
 1  4  7
 2  5  8
 3  6  9
```
In 1, 5, 9 we simply fill with a zero. In 2, 3, 6 we compute new distances, in 4, 7 and 8 we copy the values from 2, 3 and 6, respectively.

Why do that? Because it fills `r` in memory order, and the result is exactly the same.

Reply:
Ah OK this relies on the fact that the particular entry that we copy has already been set, even if that's not the case for all cells. Clever.

Reply:
Also the reason why I was confused is that the diff makes it look like the function didn't have the first loop if you don't look very carefully. :-)

Reply:
True, it took me a couple of minutes to digest the diff, too.